File: modules/virtual-people-core-serving/src/test/cc/wfa/virtual_people/core/model/conditional_assignment_impl_test.cc
// Copyright 2021 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "common_cpp/testing/common_matchers.h"
#include "common_cpp/testing/status_macros.h"
#include "common_cpp/testing/status_matchers.h"
#include "gmock/gmock.h"
#include "google/protobuf/text_format.h"
#include "gtest/gtest.h"
#include "wfa/virtual_people/common/demographic.pb.h"
#include "wfa/virtual_people/common/model.pb.h"
#include "wfa/virtual_people/core/model/attributes_updater.h"

namespace wfa_virtual_people {
namespace {

using ::wfa::EqualsProto;
using ::wfa::IsOk;
using ::wfa::StatusIs;

TEST(ConditionalAssignmentImplTest, TestNoCondition) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_assignment {
          assignments {
            source_field: "acting_demo.gender"
            target_field: "corrected_demo.gender"
          }
        }
      )pb",
      &config));
  EXPECT_THAT(AttributesUpdaterInterface::Build(config).status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ConditionalAssignmentImplTest, TestInvalidCondition) {
  // Name and value must be set for EQUAL field filter. So the build of the
  // field filter condition will fail, thus the buid of the conditional
  // assignment will fail.
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_assignment {
          condition { op: EQUAL }
          assignments {
            source_field: "acting_demo.gender"
            target_field: "corrected_demo.gender"
          }
        }
      )pb",
      &config));
  EXPECT_THAT(AttributesUpdaterInterface::Build(config).status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ConditionalAssignmentImplTest, TestEmptyAssignments) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_assignment {
          condition { op: HAS name: "acting_demo.gender" }
        }
      )pb",
      &config));
  EXPECT_THAT(AttributesUpdaterInterface::Build(config).status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ConditionalAssignmentImplTest, TestNoSourceField) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_assignment {
          condition { op: HAS name: "acting_demo.gender" }
          assignments { target_field: "corrected_demo.gender" }
        }
      )pb",
      &config));
  EXPECT_THAT(AttributesUpdaterInterface::Build(config).status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ConditionalAssignmentImplTest, TestNoTargetField) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_assignment {
          condition { op: HAS name: "acting_demo.gender" }
          assignments { source_field: "acting_demo.gender" }
        }
      )pb",
      &config));
  EXPECT_THAT(AttributesUpdaterInterface::Build(config).status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ConditionalAssignmentImplTest, TestSingleAssignment) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_assignment {
          condition { op: HAS name: "acting_demo.gender" }
          assignments {
            source_field: "acting_demo.gender"
            target_field: "corrected_demo.gender"
          }
        }
      )pb",
      &config));
  ASSERT_OK_AND_ASSIGN(std::unique_ptr<AttributesUpdaterInterface> updater,
                       AttributesUpdaterInterface::Build(config));

  // Matches the condition. Updates the event.
  LabelerEvent event_1;
  event_1.mutable_acting_demo()->set_gender(GENDER_FEMALE);
  EXPECT_THAT(updater->Update(event_1), IsOk());
  LabelerEvent expected_event_1;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        acting_demo { gender: GENDER_FEMALE }
        corrected_demo { gender: GENDER_FEMALE }
      )pb",
      &expected_event_1));
  EXPECT_THAT(event_1, EqualsProto(expected_event_1));

  // Does not match the condition. Does nothing
  LabelerEvent event_2;
  EXPECT_THAT(updater->Update(event_2), IsOk());
  EXPECT_THAT(event_2, EqualsProto(LabelerEvent()));
}

TEST(ConditionalAssignmentImplTest, TestMultipleAssignments) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_assignment {
          condition {
            op: AND
            sub_filters { op: HAS name: "acting_demo.gender" }
            sub_filters { op: HAS name: "acting_demo.age.min_age" }
            sub_filters { op: HAS name: "acting_demo.age.max_age" }
          }
          assignments {
            source_field: "acting_demo.gender"
            target_field: "corrected_demo.gender"
          }
          assignments {
            source_field: "acting_demo.age.min_age"
            target_field: "corrected_demo.age.min_age"
          }
          assignments {
            source_field: "acting_demo.age.max_age"
            target_field: "corrected_demo.age.max_age"
          }
        }
      )pb",
      &config));
  ASSERT_OK_AND_ASSIGN(std::unique_ptr<AttributesUpdaterInterface> updater,
                       AttributesUpdaterInterface::Build(config));

  // Matches the condition. Updates the event.
  LabelerEvent event_1;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        acting_demo {
          gender: GENDER_FEMALE
          age { min_age: 25 max_age: 29 }
        }
      )pb",
      &event_1));
  EXPECT_THAT(updater->Update(event_1), IsOk());
  LabelerEvent expected_event_1;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        acting_demo {
          gender: GENDER_FEMALE
          age { min_age: 25 max_age: 29 }
        }
        corrected_demo {
          gender: GENDER_FEMALE
          age { min_age: 25 max_age: 29 }
        }
      )pb",
      &expected_event_1));
  EXPECT_THAT(event_1, EqualsProto(expected_event_1));

  // Does not match the condition as acting_demo.gender is not set. Does
  // nothing.
  LabelerEvent event_2;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        acting_demo { age { min_age: 25 max_age: 29 } }
      )pb",
      &event_2));
  EXPECT_THAT(updater->Update(event_2), IsOk());
  LabelerEvent expected_event_2;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        acting_demo { age { min_age: 25 max_age: 29 } }
      )pb",
      &expected_event_2));
  EXPECT_THAT(event_2, EqualsProto(expected_event_2));

  // Does not match the condition as acting_demo.age.max_age is not set. Does
  // nothing.
  LabelerEvent event_3;
  ASSERT_TRUE(
      google::protobuf::TextFormat::ParseFromString(R"pb(
                                                      acting_demo {
                                                        gender: GENDER_FEMALE
                                                        age { min_age: 25 }
                                                      }
                                                    )pb",
                                                    &event_3));
  EXPECT_THAT(updater->Update(event_3), IsOk());
  LabelerEvent expected_event_3;
  ASSERT_TRUE(
      google::protobuf::TextFormat::ParseFromString(R"pb(
                                                      acting_demo {
                                                        gender: GENDER_FEMALE
                                                        age { min_age: 25 }
                                                      }
                                                    )pb",
                                                    &expected_event_3));
  EXPECT_THAT(event_3, EqualsProto(expected_event_3));

  // Does not match the condition as acting_demo.age.min_age is not set. Does
  // nothing.
  LabelerEvent event_4;
  ASSERT_TRUE(
      google::protobuf::TextFormat::ParseFromString(R"pb(
                                                      acting_demo {
                                                        gender: GENDER_FEMALE
                                                        age { max_age: 29 }
                                                      }
                                                    )pb",
                                                    &event_4));
  EXPECT_THAT(updater->Update(event_4), IsOk());
  LabelerEvent expected_event_4;
  ASSERT_TRUE(
      google::protobuf::TextFormat::ParseFromString(R"pb(
                                                      acting_demo {
                                                        gender: GENDER_FEMALE
                                                        age { max_age: 29 }
                                                      }
                                                    )pb",
                                                    &expected_event_4));
  EXPECT_THAT(event_4, EqualsProto(expected_event_4));
}

TEST(ConditionalAssignmentImplTest, SkipAssignmentsForUnsetField) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_assignment {
          condition { op: TRUE }
          assignments {
            source_field: "acting_demo.gender"
            target_field: "corrected_demo.gender"
          }
        }
      )pb",
      &config));
  ASSERT_OK_AND_ASSIGN(std::unique_ptr<AttributesUpdaterInterface> updater,
                       AttributesUpdaterInterface::Build(config));

  // Skip the conditional assignment as the source field is not set.
  LabelerEvent event;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        corrected_demo { gender: GENDER_FEMALE }
      )pb",
      &event));
  EXPECT_THAT(updater->Update(event), IsOk());
  LabelerEvent expected_event;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        corrected_demo { gender: GENDER_FEMALE }
      )pb",
      &expected_event));
  EXPECT_THAT(event, EqualsProto(expected_event));
}

}  // namespace
}  // namespace wfa_virtual_people

File: modules/virtual-people-core-serving/src/test/kotlin/org/wfanet/virtualpeople/core/model/ConditionalAssignmentImplTest.kt
// Copyright 2022 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.wfanet.virtualpeople.core.model

import com.google.common.truth.extensions.proto.ProtoTruth
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.wfanet.virtualpeople.common.*
import org.wfanet.virtualpeople.common.BranchNodeKt.attributesUpdater
import org.wfanet.virtualpeople.common.ConditionalAssignmentKt.assignment
import org.wfanet.virtualpeople.common.FieldFilterProto.Op

@RunWith(JUnit4::class)
class ConditionalAssignmentImplTest {

  @Test
  fun `no condition should throw`() {
    val config = attributesUpdater {
      conditionalAssignment = conditionalAssignment {
        assignments.add(
          assignment {
            sourceField = "acting_demo.gender"
            targetField = "corrected_demo.gender"
          }
        )
      }
    }
    val error = assertFailsWith<IllegalStateException> { AttributesUpdaterInterface.build(config) }
    assertTrue(error.message!!.contains("Condition is not set"))
  }

  @Test
  fun `invalid condition should throw`() {
    val config = attributesUpdater {
      conditionalAssignment = conditionalAssignment {
        condition = fieldFilterProto { op = Op.EQUAL }
        assignments.add(
          assignment {
            sourceField = "acting_demo.gender"
            targetField = "corrected_demo.gender"
          }
        )
      }
    }
    val error = assertFailsWith<IllegalStateException> { AttributesUpdaterInterface.build(config) }
    assertTrue(error.message!!.contains("Name must be set"))
  }

  @Test
  fun `empty assignment should throw`() {
    val config = attributesUpdater {
      conditionalAssignment = conditionalAssignment {
        condition = fieldFilterProto {
          op = Op.HAS
          name = "acting_demo.gender"
        }
      }
    }
    val error = assertFailsWith<IllegalStateException> { AttributesUpdaterInterface.build(config) }
    assertTrue(error.message!!.contains("No assignments"))
  }

  @Test
  fun `no source field should throw`() {
    val config = attributesUpdater {
      conditionalAssignment = conditionalAssignment {
        condition = fieldFilterProto {
          op = Op.HAS
          name = "acting_demo.gender"
        }
        assignments.add(assignment { targetField = "corrected_demo.gender" })
      }
    }
    val error = assertFailsWith<IllegalStateException> { AttributesUpdaterInterface.build(config) }
    assertTrue(error.message!!.contains("must have source_field"))
  }

  @Test
  fun `no target field should throw`() {
    val config = attributesUpdater {
      conditionalAssignment = conditionalAssignment {
        condition = fieldFilterProto {
          op = Op.HAS
          name = "acting_demo.gender"
        }
        assignments.add(assignment { sourceField = "acting_demo.gender" })
      }
    }
    val error = assertFailsWith<IllegalStateException> { AttributesUpdaterInterface.build(config) }
    assertTrue(error.message!!.contains("must have target_field"))
  }

  @Test
  fun `single assignement success`() {
    val config = attributesUpdater {
      conditionalAssignment = conditionalAssignment {
        condition = fieldFilterProto {
          op = Op.HAS
          name = "acting_demo.gender"
        }
        assignments.add(
          assignment {
            sourceField = "acting_demo.gender"
            targetField = "corrected_demo.gender"
          }
        )
      }
    }
    val updater = AttributesUpdaterInterface.build(config)

    /** Matches the condition. Updates the event. */
    val eventBuilder1 =
      labelerEvent { actingDemo = demoBucket { gender = Gender.GENDER_FEMALE } }.toBuilder()
    val expectedEvent1 = labelerEvent {
      actingDemo = demoBucket { gender = Gender.GENDER_FEMALE }
      correctedDemo = demoBucket { gender = Gender.GENDER_FEMALE }
    }
    updater.update(eventBuilder1)
    ProtoTruth.assertThat(expectedEvent1).isEqualTo(eventBuilder1.build())

    /** Does not match the condition. Does nothing. */
    val eventBuilder2 = labelerEvent {}.toBuilder()
    val expectedEvent2 = labelerEvent {}
    updater.update(eventBuilder1)
    ProtoTruth.assertThat(expectedEvent2).isEqualTo(eventBuilder2.build())
  }

  @Test
  fun `multiple assignements success`() {
    val config = attributesUpdater {
      conditionalAssignment = conditionalAssignment {
        condition = fieldFilterProto {
          op = Op.AND
          subFilters.add(
            fieldFilterProto {
              op = Op.HAS
              name = "acting_demo.gender"
            }
          )
          subFilters.add(
            fieldFilterProto {
              op = Op.HAS
              name = "acting_demo.age.min_age"
            }
          )
          subFilters.add(
            fieldFilterProto {
              op = Op.HAS
              name = "acting_demo.age.max_age"
            }
          )
        }
        assignments.add(
          assignment {
            sourceField = "acting_demo.gender"
            targetField = "corrected_demo.gender"
          }
        )
        assignments.add(
          assignment {
            sourceField = "acting_demo.age.min_age"
            targetField = "corrected_demo.age.min_age"
          }
        )
        assignments.add(
          assignment {
            sourceField = "acting_demo.age.max_age"
            targetField = "corrected_demo.age.max_age"
          }
        )
      }
    }
    val updater = AttributesUpdaterInterface.build(config)

    /** Matches the condition. Updates the event. */
    val eventBuilder1 =
      labelerEvent {
          actingDemo = demoBucket {
            gender = Gender.GENDER_FEMALE
            age = ageRange {
              minAge = 25
              maxAge = 29
            }
          }
        }
        .toBuilder()
    val expectedEvent1 = labelerEvent {
      actingDemo = demoBucket {
        gender = Gender.GENDER_FEMALE
        age = ageRange {
          minAge = 25
          maxAge = 29
        }
      }
      correctedDemo = demoBucket {
        gender = Gender.GENDER_FEMALE
        age = ageRange {
          minAge = 25
          maxAge = 29
        }
      }
    }
    updater.update(eventBuilder1)
    ProtoTruth.assertThat(expectedEvent1).isEqualTo(eventBuilder1.build())

    /** Does not match the condition as acting_demo.gender is not set. Does nothing. */
    val eventBuilder2 =
      labelerEvent {
          actingDemo = demoBucket {
            age = ageRange {
              minAge = 25
              maxAge = 29
            }
          }
        }
        .toBuilder()
    val expectedEvent2 = labelerEvent {
      actingDemo = demoBucket {
        age = ageRange {
          minAge = 25
          maxAge = 29
        }
      }
    }
    updater.update(eventBuilder2)
    ProtoTruth.assertThat(expectedEvent2).isEqualTo(eventBuilder2.build())

    /** Does not match the condition as acting_demo.age.max_age is not set. Does nothing. */
    val eventBuilder3 =
      labelerEvent {
          actingDemo = demoBucket {
            gender = Gender.GENDER_FEMALE
            age = ageRange { minAge = 25 }
          }
        }
        .toBuilder()
    val expectedEvent3 = labelerEvent {
      actingDemo = demoBucket {
        gender = Gender.GENDER_FEMALE
        age = ageRange { minAge = 25 }
      }
    }
    updater.update(eventBuilder3)
    ProtoTruth.assertThat(expectedEvent3).isEqualTo(eventBuilder3.build())

    /** Does not match the condition as acting_demo.age.min_age is not set. Does nothing. */
    val eventBuilder4 =
      labelerEvent {
          actingDemo = demoBucket {
            gender = Gender.GENDER_FEMALE
            age = ageRange { maxAge = 29 }
          }
        }
        .toBuilder()
    val expectedEvent4 = labelerEvent {
      actingDemo = demoBucket {
        gender = Gender.GENDER_FEMALE
        age = ageRange { maxAge = 29 }
      }
    }
    updater.update(eventBuilder4)
    ProtoTruth.assertThat(expectedEvent4).isEqualTo(eventBuilder4.build())
  }

  @Test
  fun `skip assignments for unset field`() {
    val config = attributesUpdater {
      conditionalAssignment = conditionalAssignment {
        condition = fieldFilterProto {
          op = Op.HAS
          name = "acting_demo.age"
        }
        assignments.add(
          assignment {
            sourceField = "acting_demo.age.min_age"
            targetField = "corrected_demo.age.min_age"
          }
        )
        assignments.add(
          assignment {
            sourceField = "acting_demo.age.max_age"
            targetField = "corrected_demo.age.max_age"
          }
        )
      }
    }
    val updater = AttributesUpdaterInterface.build(config)

    /** Skip the conditional assignment on min_age as the source field is not set. */
    val eventBuilder1 =
      labelerEvent { actingDemo = demoBucket { age = ageRange { maxAge = 29 } } }.toBuilder()
    val expectedEvent1 = labelerEvent {
      actingDemo = demoBucket { age = ageRange { maxAge = 29 } }
      correctedDemo = demoBucket { age = ageRange { maxAge = 29 } }
    }
    updater.update(eventBuilder1)
    ProtoTruth.assertThat(expectedEvent1).isEqualTo(eventBuilder1.build())
  }
}
